### ISSUE TITLE
Proposal: Conditional rules based on the environment

### ISSUE THREAD

**Issue by AnilRedshift** [eslint/eslint#6460]
**What version of ESLint are you using?**
2.13.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

Hello, I have a scenario where I would like to have different .eslint rules when developing locally on my machine than when it's time to submit the project. 
Here is how eslint is currently useful to me for local development:
As part of the build process, we run something like (simplified): `rm -rf dist/ && eslint app/js && browserify app/js/index.js`

The nice thing about this setup is that the build fails if eslint doesn't pass, so I don't spend time debugging or testing code that is bad in some way. This is mostly goodness, except for when I want to debug something using banned rules. A couple of classic examples are `no-constant-condition` and `no-console`. What I want is to be able to say: Have two configs: The main config, and then a few modifications for local development.

Currently, as far as I can tell, I have to have separate config files and pass them in the command line. This is sub-optimal because the rules are in different places and it is not easy to extend off a rule. As far as I know (I may be wrong), I have to essentially copy/paste the .eslintrc and then make the modifications I want for the dev version.

My proposal is to add conditional rules in the form of env:[your environment here]/[rule]
For example, I would create this .eslintrc:
```
no-constant-condition: ["error"],
env:dev/no-constant-condition: ["warn"],
```
Which would allow me to place both variants side-by-side and then to trigger the code would be `eslint --env dev ` which is easily configurable by build scripts.

I have a local modification to eslint which accomplishes this, and I am happy to drive the fix to mainline if folks like the proposal. This is my first time working with this project so I would appreciate patience and help getting it in.

**Comment by platinumazure** [comment]
Unfortunately, the term "env"/"environment" is overloaded here- currently an environment refers to an execution context, such as writing browser code or using QUnit globals. So we would probably need to come up with another term for your proposal.

Otherwise, this seems intriguing and I would love to see a robust discussion on it. That said, you yourself have acknowledged this can already be done using multiple config files, so we'll see how far this goes.

**Comment by ilyavolodin** [comment]
I really don't think that ESLint should be aware of where it's executing from. It shouldn't care. It's the job of the environment to set itself up in a way where the right thing gets executed. ESlint already have ability for you to specify which file to use, so you can just change the script that you run on the server to this:

```sh
rm -rf dist/ && eslint app/js -c .eslintrcProd && browserify app/js/index.js
```

Also this is a reason why we added ability to set rules as warnings, and not errors. You can have a single config file, and mark the rules that you don't want to break prod build as warnings. This way you still get notify about them while you are developing, but they will not break the build.

Overall I'm very much :-1: on this.

**Comment by nzakas** [comment]
If we ignore the specific suggestion and look at it from a higher level, what's being requested is the ability to conditionally apply rules, so it's similar in spirit to the glob-based config discussion. I wonder if we should be considering a general approach to condition-based overrides that could support globs as well as some other conditions?

**Comment by hzoo** [comment]
Babel has a similar concept: http://babeljs.io/docs/usage/babelrc/#env-option, but yeah you could make different .eslintrc files for this

**Comment by ilyavolodin** [comment]
@nzakas I think this is opposite to glob-based config. Glob-based config is for selectively choosing files that would be linted. This is for linting all files with selective configuration.
I'm not sure generic condition-based logic is something we want to get into. That's what task runners are for.

**Comment by platinumazure** [comment]
@ilyavolodin I disagree. Glob based config should be broader than that. For
example, it should be usable for triggering a plugin processor based on
file extension.

**Comment by nzakas** [comment]
@ilyavolodin I think you're thinking too narrowly about this. Glob-based matching says, "for these files, here's a different config." My point is that "for these files" can be abstracted out to "for this condition", so there's little difference between "for these files" and "for this environment." So the word "opposite" is incorrect here. 

And I'm not arguing for this proposal, as I think there are problems with it (not the least if which being the use of the word "environment"), but I am arguing that we should think about condition-based configs as a class of thing to support rather than saying we will support one condition but no others.

**Comment by AnilRedshift** [comment]
Thanks for all of the input. I am terrible with naming things - consider it just a placeholder for the discussion we're having now.

@ilyavolodin With a separate config file (.eslintrcProd), do you have any suggestions on having to effectively copy/paste the file twice? The only supported way that I can see would be to build the whole "base configuration" npm package and then riff off of it.

An alternate proposal that would solve my pain point would be to do just that - build in support for `extends` such that it could extend off a local file. The spec would probably have to be updated to differentiate from an NPM package but that could work as well.

Thanks!

**Comment by ilyavolodin** [comment]
@AnilRedshift you actually can `extend` from [local files](http://eslint.org/docs/user-guide/configuring#using-a-configuration-file). And that's what I would suggest to do. Create base config file, then create 2 config files per runtime environment and extend base in each one of those. Run ESLint with `-c` flag. Or just create a single config file and mark all of the rules that you don't want to fail prod build as "warning" instead of "error".

**Comment by AnilRedshift** [comment]
Cool, that's enough to work around my needs. IMO baking some of this logic into the eslint files feels more righteous in that the logic as pertains to eslint **stays** with eslint but I can see why folks would want to avoid feature creep as well.

If it were up to me, my test would be: What makes eslint the easiest to use for our users that is part of our core mission? IMO being able to configure a dev environment is part of that core mission.

**Comment by alberto** [comment]
Thanks for your interest in improving eslint. Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
